private/bufpkg/bufprotoplugin/cmd/protoc-gen-go-apiclientconnect/main.go

```diff
@@ -0,0 +1,424 @@
+	g.P(`// buildAddress modifies the given address with any additional options for transport such as the scheme and any subdomains`)
+	g.P(`func (p *provider) buildAddress(address string) string {`)
+	g.P(`if p.addressMapper != nil {`)
+	g.P(`address = p.addressMapper(address)`)
+	g.P(`}`)
+	g.P(`if p.scheme != 0 {`)
+	g.P(`address = p.scheme.String()+"://"+address`)
+	g.P(`}`)
+	g.P(`return address`)
+	g.P(`}`)
+	g.P()
+
+	// Import path for the api named_go_package
+	apiGoImportPath, err := helper.NewPackageGoImportPath(
+		goPackageFileSet,
+		"api",
+	)
+	if err != nil {
+		return err
+	}
+
+	// Iterate over the services and create a constructor function for each
+	for _, service := range goPackageFileSet.Services() {
+		interfaceName := service.GoName
+		interfaceGoIdent := apiGoImportPath.Ident(interfaceName)
+		interfaceGoIdentString := g.QualifiedGoIdent(interfaceGoIdent)
+
+		g.P(`func (p *provider) New`, interfaceName, `(ctx `, contextGoIdentString, `, address string) (`, interfaceGoIdentString, `, error) {`)
+		g.P(`var contextModifier func(context.Context) context.Context`)
+		g.P(`var err error`)
+		g.P(`if p.contextModifierProvider != nil {`)
+		g.P(`contextModifier, err = p.contextModifierProvider(address)`)
+		g.P(`if err != nil {`)
+		g.P(`return  nil, err`)
+		g.P(`}`)
+		g.P(`}`)
+
+		g.P(`return New`, interfaceName, `Client(`)
+		g.P(`p.httpClient,`)
+		g.P(`p.buildAddress(address),`)
+		g.P(`contextModifier,`)
+		g.P(withGRPCIdentString, `(),`)
+		g.P(`), nil`)
+		g.P(`}`)
+	}
+
+	return nil
+}
+
+func generateServiceFile(helper protogenutil.NamedHelper, plugin *protogen.Plugin, file *protogen.File) error {
+	g, err := helper.NewGeneratedFile(plugin, file, pluginName)
+	if err != nil {
+		return err
+	}
+	contextGoIdentString := g.QualifiedGoIdent(contextPackage.Ident("Context"))
+	loggerGoIdentString := g.QualifiedGoIdent(zapPackage.Ident("Logger"))
+
+	// Import path for the connect named_go_package
+	connectGoImportPath, err := helper.NewGoImportPath(
+		file,
+		"connect",
+	)
+	if err != nil {
+		return err
+	}
+	httpClientGoIdentString := g.QualifiedGoIdent(connectGoPackage.Ident("HTTPClient"))
+
+	for _, service := range file.Services {
+		interfaceName := service.GoName
+		structName := protogenutil.GetUnexportGoName(interfaceName)
+		clientGoIdent := connectGoImportPath.Ident(interfaceName + "Client")
+		clientGoIdentString := g.QualifiedGoIdent(clientGoIdent)
+		newClientGoIdent := connectGoImportPath.Ident("New" + interfaceName + "Client")
+		newClientGoIdentString := g.QualifiedGoIdent(newClientGoIdent)
+		newRequestGoIdent := connectGoPackage.Ident("NewRequest")
+		newRequestGoIdentString := g.QualifiedGoIdent(newRequestGoIdent)
+
+		g.P(`type `, structName, `Client struct {`)
+		g.P(`logger *`, loggerGoIdentString)
+		g.P(`client `, clientGoIdentString)
+		g.P(`contextModifier func (`, contextGoIdentString, `) `, contextGoIdentString)
+		g.P(`}`)
+		g.P()
+
+		g.P(`func New`, interfaceName, `Client(`)
```

Comment:
By exposing this constructor function, we now have the `contextModifier` in the arg list.  This isn't used by the callers in `buf/core`, but they should be able to pass `nil` for this value since the generated code for the service files checks whether the `contextModifier` is nil before applying it to the context.

Reply:
I don't love it. But I think this is an acceptable trade-off that allows us to instantiate new connect clients that implement the `registryv1alpha1api` set of interfaces generated by protoc-gen-go-api.

This enables us to use single clients (which we do elsewhere) without the provider abstraction. 

The alternative is to unexport the constructors and be explicit about the clients we pass around. Although this does mean always having to instantiate a provider via `NewProvider`.

I'll leave this to your discretion.

Reply:
Just a drive-by comment, but throughout buf, constructor arguments are required, optional arguments are options...not sure if relevant here yet

Reply:
I think @mfridman and I figured out the point of confusion, tldr we should choose one of `Provider.New*` or `New*`